### PR TITLE
fixing docstring styles in crop module (2/3)

### PR DESCRIPTION
**Pull request by matthew7838**
### Summary
Fixing the crop file's docstring to the correct format. There will be one additional PR since the changes exceed expectations.

### CODE REVIEW

File: RUFAS/routines/field/crop/heat_units.py [hunk 1]
```diff
@@ -1,28 +1,55 @@
...
-            min_air_temperature: minimum air temperature for the day (C)
-            max_air_temperature: maximum air temperature for the day (C)
-
-        SWAT References: 5:1.1, 5:2.1.2,
+        """
+        Main function for absorbing heat units during a day and accumulating them.
+
+        Parameters
+        ----------
+        mean_air_temperature : float
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/heat_units.py]:
```suggestion
        mean_air_temperature : Optional[float]
```

File: RUFAS/routines/field/crop/heat_units.py [hunk 2]
```diff
@@ -42,52 +69,75 @@ def absorb_heat_units(self, mean_air_temperature: float = None,
...
         if self.data.use_heat_unit_temperature or (air_temperature is None):  # alternative method
             self.data.new_heat_units = self._determine_new_heat_units(self.data.heat_unit_temperature,
                                                                       self.data.minimum_temperature)
         else:  # main method
             self.data.new_heat_units = self._determine_new_heat_units(air_temperature, self.data.minimum_temperature)
 
     def add_heat_units(self) -> None:
-        """add newly acquired heat units to accumulated heat units"""
+        """
+        add newly acquired heat units to accumulated heat units
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/heat_units.py]:
```suggestion
        Add newly acquired heat units to accumulated heat units.
```

File: RUFAS/routines/field/crop/heat_units.py [hunk 3]
```diff
@@ -42,52 +69,75 @@ def absorb_heat_units(self, mean_air_temperature: float = None,
...
+
+        Parameters
+        ----------
+        mean_air_temperature : Optional[float], optional
+            Average air temperature for the day (°C).
+        min_air_temperature : Optional[float], optional
+            Minimum air temperature for the day (°C).
+        max_air_temperature : Optional[float], optional
+            Maximum air temperature for the day (°C).
+
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/heat_units.py:140]: Needs a "Raises" section here.

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 4]
```diff
@@ -3,17 +3,30 @@
...
+        Reference to the provided `CropData` instance or a new default instance.
+
+    """
     def __init__(self, crop_data: Optional[CropData] = None):
         self.data = crop_data or CropData()  # initialize with defaults, if not given
 
     def grow_canopy(self) -> None:
-        """main leaf area index function"""
+        """
+        main leaf area index function
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
        Main leaf area index function.
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 5]
```diff
@@ -42,34 +55,79 @@ def grow_canopy(self) -> None:
...
 
     def determine_leaf_area_added(self) -> None:
-        """sets actual leaf area added, by adjusting for the plant growth factor
-        SWAT Reference: 5:3.2.2
+        """
+        Sets the actual leaf area added, adjusted for the plant growth factor.
+
+        References
+        ----------
+        SWAT :3.2.2
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
        SWAT 5:3.2.2
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 6]
```diff
@@ -42,34 +55,79 @@ def grow_canopy(self) -> None:
...
+
+        Returns
+        -------
+        float
+            The current height of the canopy (meters).
+
+        References
+        ----------
+        SWAT 5:2.1.14
+
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py:129]: Needs a "Raises" section

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 7]
```diff
@@ -109,66 +184,104 @@ def _determine_lai_shapes(first_heat_fraction: float, second_heat_fraction: floa
...
+        Parameters
+        ----------
+        leaf_area_fraction : float
+            Optimal leaf area fraction for the day (uniteless).
+        previous_leaf_area_fraction : float
+            Previous day's optimal leaf area fraction (uniteless).
+        max_leaf_area_index : float
+            The maximum leaf area index achievable by the plant (uniteless).
+        previous_leaf_area_index : float
+            The previous day's leaf area index (uniteless).
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
        leaf_area_fraction : float
            Optimal leaf area fraction for the day (unitless).
        previous_leaf_area_fraction : float
            Previous day's optimal leaf area fraction (unitless).
        max_leaf_area_index : float
            The maximum leaf area index achievable by the plant (unitless).
        previous_leaf_area_index : float
            The previous day's leaf area index (unitless).
```

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 8]
```diff
@@ -87,50 +117,80 @@ def uptake_nitrogen(self, layer_nitrates: List[float], layer_depths: List[float]
 
     # ---- member functions (setters, internal utility, call sub-routines) ----
     def shift_nitrogen_time(self) -> None:
-        """copies the current nitrogen value to previous_nitrogen (for use between time steps)"""
+        """
+        copies the current nitrogen value to previous_nitrogen (for use between time steps)
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py]:
```suggestion
        Copies the current nitrogen value to previous_nitrogen (for use between time steps).
```

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 9]
```diff
@@ -182,24 +266,38 @@ def fix_nitrogen(self, water_factor: float) -> None:
...
+        Notes
+        -----
+        SWAT assumes that the difference between the nutrient fraction near maturity and the nutrient fraction at
+        maturity in the crop is equal to 0.00001 (as per SWAT theoretical documentation pages 331 and 336, top
+        paragraphs of both). Therefore, the near mature nutrient fraction is adjusted to meet that assumption in this
+        calculation.
+
+        References
+        ----------
+        SWAT 5:2.3.2, 5:2.3.3, 5:2.3.20, 5:2.3.21
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py:300]: Needs a "Raises" section.

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 10]
```diff
@@ -229,17 +327,29 @@ def determine_nutrient_shape_parameters(half_mature_heat_fraction: float, mature
...
-        SWAT Reference: Equations 5:2.3.2, 5:2.3.3, 5:2.3.20, 5:2.3.21
+        Returns
+        -------
+        float
+            The logarithmic term of the nitrogen shape coefficients, crucial for calculating the shape coefficients
+            used in nitrogen uptake modeling (unitless).
+
+        References
+        ----------
+        SWAT 5:2.3.2, 5:2.3.3, 5:2.3.20, 5:2.3.21
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py:368]: Needs a "Raises" section.

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 11]
```diff
@@ -396,17 +561,29 @@ def determine_layer_nutrient_uptake_potential(layer_bounds: List[float], total_d
...
+
+        Returns
+        -------
+        float
+            The potential amount of nitrogen that can be taken up from the soil surface to the specified depth (kg/ha).
 
-        SWAT Reference: Equations 5:2.3.6, 5:2.3.24
+        References
+        ----------
+        SWAT 5:2.3.6, 5:2.3.24
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py:613]: Needs a "Raises" section

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Docstrings look good, but there are missing "Raises" sections. Please add these and make sure they match the RuFaS "Raises" style outlined in #1074.

File: RUFAS/routines/field/crop/heat_units.py [hunk 12]
```diff
@@ -1,28 +1,55 @@
...
+        mean_air_temperature : Optional[float]
+            Average air temperature for the day (°C).
+        min_air_temperature : Optional[float]
+            Minimum air temperature for the day (°C).
+        max_air_temperature : Optional[float]
+            Maximum air temperature for the day (°C).
+
+        Notes
+        -----
+        If the attribute `use_heat_unit_temperature` is False, both `min_air_temperature` and `max_air_temperature` are
```
Comment (allisterakun) [inline RUFAS/routines/field/crop/heat_units.py]: Maybe clarify that the attribute `use_heat_unit_temperature` is from the CropData class/self.data

**Review by allisterakun — APPROVED** [review]
Good job, Matthew!

File: RUFAS/routines/field/crop/heat_units.py [hunk 13]
```diff
@@ -42,52 +69,83 @@ def absorb_heat_units(self, mean_air_temperature: float = None,
...
+        max_air_temperature : Optional[float], optional
+            Maximum air temperature for the day (°C).
+
+        Raises
+        ------
+        ValueError
+            If `use_heat_unit_temperature` is True, both `min_air_temperature` and `max_air_temperature` must be
+            provided.
+        ValueError
+            If `use_heat_unit_temperature` is False, `mean_air_temperature` must be provided.
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/heat_units.py]: This does not follow the style of "Raises" sections in #1074.
```suggestion
        ValueError
            If `use_heat_unit_temperature` is True and both `min_air_temperature` and `max_air_temperature` are not provided.
            If `use_heat_unit_temperature` is False and `mean_air_temperature` is not provided.
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 14]
```diff
@@ -42,34 +55,86 @@ def grow_canopy(self) -> None:
...
+        References
+        ----------
+        SWAT 5:2.1.14
+
+        Raises
+        ------
+        ValueError
+            Max_canopy_height is provided.
+        ValueError
+            optimal_leaf_area_fraction is not a valid fraction.
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]: These descriptions do not accurately reflect what causes ValueErrors to be raised.

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 15]
```diff
@@ -182,24 +266,43 @@ def fix_nitrogen(self, water_factor: float) -> None:
...
+        calculation.
+
+        References
+        ----------
+        SWAT 5:2.3.2, 5:2.3.3, 5:2.3.20, 5:2.3.21
+
+        Raises
+        ------
+        ValueError
+            half_mature_heat_fraction equals mature_heat_fraction.
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py]:
```suggestion
            If half_mature_heat_fraction equals mature_heat_fraction.
```

File: RUFAS/routines/field/crop/nitrogen_incorporation.py [hunk 16]
```diff
@@ -363,17 +538,41 @@ def determine_deepest_accessible_layer(root_depth: float, layer_bounds: List[flo
...
+            If there are duplicate depths, indicating multiple soil layers at the same depth.
 
-        Args:
-            layer_bounds: list of lower boundaries for each soil layer, in ascending order (i.e., increasing depths)
-            total_demand: total nitrogen demand of the plant
-            root_depth: current depth of the plant roots
-            nutrient_distribution_parameter: nitrogen uptake distribution parameter
+        References
+        ----------
+            pseudocode: C.5.C.2, C.5.C.3
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/nitrogen_incorporation.py]:
```suggestion
        ----------
        pseudocode: C.5.C.2, C.5.C.3
```

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
The style of "Raises" sections should adhere to what is outlined in #1074.

**Review by Isgotkowitz — APPROVED** [review]
LGTM, thanks for working through those changes Matthew

**Pull request merged**